odoo/tools/profiler.py

```diff
@@ -272,6 +272,94 @@ def post_process(self):
         super().post_process()
 
 
+class QwebCollector(Collector):
+    """
+    Record qweb execution with directive trace.
+    """
+    name = 'qweb'
+
+    def __init__(self):
+        super().__init__()
+        self.archs = {}
+        self.results = []
+        self.stack = []
+
+        def hook(event, env, **kwargs):
+            if event == 'begin':
+                self._handle_begin(env, **kwargs)
+            elif event == 'end':
+                self._handle_end(env, **kwargs)
+        self.hook = hook
+
+    def _handle_begin(self, env, view_id, xpath, directive, attrib, arch):
+        self.archs[view_id] = arch
+        self.current = {
+            'view_id': view_id,
+            'xpath': xpath,
+            'directive': self._get_directive_profiling_name(directive, attrib),
+            'delay': 0,
+            'query': 0,
+            'now': (time.time(), env.cr.sql_log_count),
+        }
+        self.results.append(self.current)
+        self.stack.append(self.current)
+
+    def _handle_end(self, env):
+        data = self.stack.pop()
+        data['delay'] = time.time() - data['now'][0]
+        data['query'] = env.cr.sql_log_count - data['now'][1]
+
+    def _get_directive_profiling_name(self, directive, attrib):
+        expr = ''
+        if directive == 'set':
+            expr = f'''t-set={repr(attrib['t-set'])}'''
+            if 't-value' in attrib:
+                expr = f'''{expr} t-value={repr(attrib['t-value'])}'''
+            if 't-valuef' in attrib:
+                expr = f'''{expr} t-valuef={repr(attrib['t-valuef'])}'''
+        elif directive == 'foreach':
+            expr = f'''t-foreach={repr(attrib['t-foreach'])} t-as="{repr(attrib['t-as'])}'''
```

Comment:
The single  `"` here is strange. I would remove it and replace all `'''` with `"`